nemo/collections/nlp/models/language_modeling/megatron_base_prompt_learning_model.py

```diff
@@ -0,0 +1,502 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import re
+from collections import OrderedDict
+
+import torch
+from omegaconf.dictconfig import DictConfig
+from omegaconf.omegaconf import open_dict
+from pytorch_lightning.trainer.trainer import Trainer
+from torch import Tensor
+
+from nemo.collections.nlp.models.language_modeling.megatron_base_model import MegatronBaseModel
+from nemo.collections.nlp.modules.common import (
+    PromptEncoder,
+    PromptTable,
+    VirtualPromptPlaceholderToken,
+    VirtualPromptSource,
+    VirtualPromptStyle,
+)
+from nemo.collections.nlp.modules.common.transformer.text_generation import TextGeneration
+from nemo.utils import logging
+
+__all__ = ['MegatronBasePromptLearningModel']
+
+
+class MegatronBasePromptLearningModel(MegatronBaseModel, TextGeneration):
+    """
+    Model class for prompt-tuning or p-tuning a pretrained Megatron model. 
+
+    Prompt Tuning initalizes virtual prompt embeddings directly from a copy of
+    certain token embeddings from the the pretrained model's vocabulary
+    and directly tunes these embedding weights. The token embeddings used in 
+    initalization are specified by the user in the config file. The model can 
+    be prompt-tuned for multiple tasks at once. virtual prompts are stored in a 
+    prompt table and can be added or deleted without disrupting virtual prompts 
+    for other tasks. 
+
+    P-tuning initializes an LSTM encoder model that generates virtual prompt
+    embeddings for every task. Each task shares the same encoder. After ptuning
+    is compelete, the learned virtual prompts can be saved to the prompt table
+    using add_ptuned_prompts_to_prompt_table(). Thus, if a user wants to add a 
+    new virtual prompt via p-tuning, they do not need to retrain on all previous 
+    tasks. This gives p-tuning the same task flexiblity as prompt-tuning.
+    """
+
+    def __init__(self, cfg: DictConfig, trainer: Trainer):
+        super().__init__(cfg, trainer)
+
+        self.cfg = cfg
+
+        self.load_frozen_model(cfg, trainer)
+
+        self.tokenizer = self.frozen_model.tokenizer
+
+        if self.frozen_model.cfg.precision == 16:
```

Comment:
You should be able to get this from `frozen_model.autocast_dtype` instead of this if-else.